File: api/app/pricing.py
"""Server-side pricing.

Mirrors the shape of the web-side estimator but takes its weight & time
inputs from the real slicer when available. Pricing logic must live server-
side so the client can't forge a cheaper quote.
"""
from __future__ import annotations

from dataclasses import dataclass, asdict


# Keep in step with src/lib/catalog.ts — MATERIAL_DENSITY_G_PER_CM3 and
# MATERIAL_RATE_GBP_PER_GRAM. The TS side is canonical (it's what the
# creator sees first); these exist because pricing must also be computed
# somewhere the client can't forge.
MATERIAL_DENSITY = {
    "PLA": 1.24,
    "PETG": 1.27,
    "ABS": 1.04,
    "ASA": 1.07,
    "TPU": 1.21,
}
MATERIAL_PRICE_PER_G = {
    "PLA": 0.045,
    "PETG": 0.058,
    "ABS": 0.064,
    "TPU": 0.095,
}

# Shell + infill estimation, mirroring src/lib/filament.ts. Only used when
# PrusaSlicer is unavailable — when it runs, weight comes from the G-code.
WALL_THICKNESS_MM = 0.9
WASTE_FACTOR = 1.07


def estimate_printed_volume_cm3(
    *,
    volume_cm3: float,
    surface_area_cm2: float,
    infill_pct: int,
    wall_thickness_mm: float = WALL_THICKNESS_MM,
) -> float:
    """Volume of plastic actually extruded for a hollow FDM part."""
    if volume_cm3 <= 0:
        return 0.0
    area = surface_area_cm2 if surface_area_cm2 > 0 else 6 * (volume_cm3 ** (2 / 3))
    shell_cm3 = min(area * (wall_thickness_mm / 10.0), volume_cm3)
    interior_cm3 = volume_cm3 - shell_cm3
    fill = min(1.0, max(0.0, infill_pct / 100.0))
    return shell_cm3 + interior_cm3 * fill

QUALITY_LAYER_MM = {"draft": 0.28, "standard": 0.20, "fine": 0.12}
QUALITY_TIME_MULT = {"draft": 0.7, "standard": 1.0, "fine": 1.7}

DELIVERY_FEE = {"pickup": 0.0, "courier": 7.5}

SERVICE_FEE = 1.5
MARGIN = 1.4
MACHINE_RATE_PER_HOUR = 2.4
COLOR_CHANGE_SURCHARGE = 0.4  # GBP per extra colour for purge / waste tower


@dataclass
class Quote:
    weight_g: float
    time_minutes: float
    material_cost: float
    machine_cost: float
    service_fee: float
    delivery: float
    subtotal: float
    discount_applied: float
    multi_material_surcharge: float
    total: float
    engine: str  # "prusa-slicer" | "volume-estimate"


def quote(
    *,
    volume_cm3: float,
    material: str,
    quality: str,
    infill_pct: int,
    quantity: int,
    delivery: str,
    surface_area_cm2: float = 0.0,
    weight_g_override: float | None = None,
    time_minutes_override: float | None = None,
    engine: str = "volume-estimate",
    discount_pct: int = 0,
    free_mode: bool = False,
    color_count: int = 1,
    price_per_g_override: float | None = None,
) -> Quote:
    density = MATERIAL_DENSITY[material]
    # Maker-set filament rates will arrive here as an override; the
    # platform default is only the fallback.
    price_per_g = (
        price_per_g_override
        if price_per_g_override is not None
        else MATERIAL_PRICE_PER_G[material]
    )
    time_mult = QUALITY_TIME_MULT[quality]

    if weight_g_override is not None:
        weight_g = weight_g_override * quantity
    else:
        printed_cm3 = estimate_printed_volume_cm3(
            volume_cm3=volume_cm3,
            surface_area_cm2=surface_area_cm2,
            infill_pct=infill_pct,
        )
        weight_g = printed_cm3 * density * WASTE_FACTOR * quantity

    if time_minutes_override is not None:
        est_minutes = time_minutes_override * quantity
    else:
        infill_factor = 0.25 + (infill_pct / 100.0) * 0.75
        per_part_minutes = volume_cm3 * 1.6 * time_mult * (0.6 + infill_factor * 0.4)
        est_minutes = per_part_minutes * quantity

    material_cost = weight_g * price_per_g
    machine_cost = (est_minutes / 60.0) * MACHINE_RATE_PER_HOUR
    list_subtotal = (material_cost + machine_cost) * MARGIN
    delivery_fee = DELIVERY_FEE[delivery]

    if free_mode:
        subtotal = 0.0
    else:
        d = max(0, min(100, int(discount_pct)))
        subtotal = list_subtotal * (1 - d / 100.0)

    discount_applied = list_subtotal - subtotal

    extra_colors = max(0, int(color_count) - 1)
    multi_material_surcharge = (
        0.0 if free_mode else extra_colors * COLOR_CHANGE_SURCHARGE
    )

    total = subtotal + SERVICE_FEE + delivery_fee + multi_material_surcharge

    return Quote(
        weight_g=round(weight_g, 2),
        time_minutes=round(est_minutes, 1),
        material_cost=round(material_cost, 2),
        machine_cost=round(machine_cost, 2),
        service_fee=SERVICE_FEE,
        delivery=round(delivery_fee, 2),
        subtotal=round(subtotal, 2),
        discount_applied=round(discount_applied, 2),
        multi_material_surcharge=round(multi_material_surcharge, 2),
        total=round(total, 2),
        engine=engine,
    )


def as_dict(q: Quote) -> dict:
    return asdict(q)

File: src/lib/pricing.ts
import {
  DELIVERY_OPTIONS,
  MACHINE_TIME_MIN_GBP,
  MACHINE_TIME_RATE_GBP_PER_HOUR,
  MARGIN_MULTIPLIER,
  MATERIALS,
  QUALITIES,
  SERVICE_FEE_BASE_GBP,
  SERVICE_FEE_PCT,
  quantityTierDiscountPct,
  type MaterialKey,
  type QualityKey,
} from "./catalog";
import { isPlatformFeePromoActive } from "./promotions";
import {
  DEFAULT_INFILL_FRACTION,
  estimateFilamentGrams,
  materialCostGbp,
} from "./filament";

/** Industry-rough purge / waste-tower cost per extra colour change. */
export const COLOR_CHANGE_SURCHARGE_GBP = 0.4;

export type Quote = {
  weightG: number;
  estMinutes: number;
  materialCost: number;
  machineCost: number;
  /** Actually charged service fee (0 during the launch promo, on
   *  free-mode jobs, or when the affiliate creator-waiver fires). */
  serviceFee: number;
  /** What the service fee would be without any waiver — for the
   *  strikethrough display in the breakdown. Equal to serviceFee when
   *  no waiver is active. */
  serviceFeeListPrice: number;
  /** Launch promo. */
  promoApplied: boolean;
  /** Affiliate creator-waiver — referred creator's first paid job. */
  affiliateWaiverApplied: boolean;
  delivery: number;
  subtotal: number;
  discountApplied: number; // £ saved vs list price (before service fee)
  /** Volume-break % actually applied (0 when none, or when a bigger
   *  community discount won). Drives the "10+ −10%" badge. */
  quantityTierPct: number;
  multiMaterialSurcharge: number; // £ added for purge / colour changes
  total: number;
};

/**
 * Stage-1 estimate: shell + infill from the mesh's own geometry (see
 * `./filament`). Stage-2 replaces `weightG` and `estMinutes` with values
 * from the server-side slicer.
 *
 * Community discount semantics:
 *   - `freeMode` wins → subtotal = 0 (service fee still applies; delivery too)
 *   - `discountPct` applies to the (material + machine) * margin subtotal
 *   - Service fee + delivery are never discounted
 */
export function estimateQuote({
  volumeCm3,
  surfaceAreaCm2 = 0,
  material,
  quality,
  infillPct,
  quantity,
  delivery,
  discountPct = 0,
  freeMode = false,
  colorCount = 1,
  deliveryFeeOverride,
  creatorReferralEligible = false,
  weightGPerPartOverride,
  ratePerGramGbpOverride,
}: {
  volumeCm3: number;
  /** Mesh triangle area. 0 falls back to a cube approximation — see
   *  `approxSurfaceAreaCm2`. */
  surfaceAreaCm2?: number;
  material: MaterialKey;
  quality: QualityKey;
  infillPct: number;
  quantity: number;
  delivery: "pickup" | "courier";
  discountPct?: number;
  freeMode?: boolean;
  /** Number of distinct colours in the print. >1 triggers AMS surcharge. */
  colorCount?: number;
  /** Override the static catalogue delivery fee with a live courier quote. */
  deliveryFeeOverride?: number;
  /** When true, the creator is a referred user on their first paid job —
   *  the service fee is dropped to 0 (the maker-side cut still fires
   *  and gets redirected to the affiliate at capture time). */
  creatorReferralEligible?: boolean;
  /** Known filament mass for ONE unit — from the slicer, or from a design
   *  job's recorded metrics. Skips the geometric estimate entirely. */
  weightGPerPartOverride?: number;
  /** Maker's own filament rate (£/g) once makers set one. Falls back to
   *  the platform default from the catalogue. */
  ratePerGramGbpOverride?: number;
}): Quote {
  const mat = MATERIALS.find((m) => m.key === material)!;
  const q = QUALITIES.find((qq) => qq.key === quality)!;

  const infillFraction =
    infillPct > 0 ? infillPct / 100 : DEFAULT_INFILL_FRACTION;

  const perPartWeightG =
    weightGPerPartOverride ??
    estimateFilamentGrams({
      volumeCm3,
      surfaceAreaCm2,
      infillFraction,
      densityGPerCm3: mat.densityGPerCm3,
    });
  const weightG = perPartWeightG * quantity;

  // Time is still the old volume heuristic — the slicer's own estimate
  // replaces it at stage 2, which is the only way to get it honest.
  const timeInfillFactor = 0.25 + infillFraction * 0.75;
  const perPartMinutes =
    volumeCm3 * 1.6 * q.timeMultiplier * (0.6 + timeInfillFactor * 0.4);
  const estMinutes = perPartMinutes * quantity;

  const ratePerGram = ratePerGramGbpOverride ?? mat.pricePerGramGbp;
  const materialCost = materialCostGbp(weightG, ratePerGram);
  const rawMachineCost = (estMinutes / 60) * MACHINE_TIME_RATE_GBP_PER_HOUR;
  const machineCost = Math.max(MACHINE_TIME_MIN_GBP, rawMachineCost);
  const deliveryFee =
    deliveryFeeOverride !== undefined
      ? deliveryFeeOverride
      : (DELIVERY_OPTIONS.find((d) => d.key === delivery)?.priceGbp ?? 0);

  const listSubtotal = (materialCost + machineCost) * MARGIN_MULTIPLIER;

  // Volume break for multi-unit runs. It does NOT stack with a community
  // discount — the creator gets whichever is larger, so the two schemes can
  // be tuned independently without compounding toward a free print.
  const tierPct = quantityTierDiscountPct(quantity);
  const effectivePct = Math.max(discountPct, tierPct);
  const quantityTierPct = tierPct > discountPct ? tierPct : 0;

  let subtotal = listSubtotal;
  if (freeMode) subtotal = 0;
  else if (effectivePct > 0)
    subtotal = listSubtotal * (1 - Math.min(100, Math.max(0, effectivePct)) / 100);

  const discountApplied = listSubtotal - subtotal;
  // Service fee = £2 base + 10% of the printing subtotal. Waivers that
  // bring it to 0:
  //   - free-mode / £0 subtotal (no money changes hands)
  //   - launch promo
  //   - affiliate creator-waiver on a referred user's first paid job
  // List price is still computed for the strikethrough display.
  const serviceFeeListPrice = SERVICE_FEE_BASE_GBP + subtotal * SERVICE_FEE_PCT;
  const promoApplied = isPlatformFeePromoActive();
  const freeJob = subtotal === 0;
  const affiliateWaiverApplied = creatorReferralEligible && !freeJob;
  const serviceFee =
    promoApplied || freeJob || affiliateWaiverApplied ? 0 : serviceFeeListPrice;

  // Per-extra-colour purge surcharge. Applied per part (not multiplied by
  // quantity since AMS sequences colours within a single multi-part job).
  const extraColors = Math.max(0, colorCount - 1);
  const multiMaterialSurcharge = freeMode
    ? 0
    : extraColors * COLOR_CHANGE_SURCHARGE_GBP;

  const total = subtotal + serviceFee + deliveryFee + multiMaterialSurcharge;

  return {
    weightG,
    estMinutes,
    materialCost,
    machineCost,
    serviceFee,
    serviceFeeListPrice,
    promoApplied,
    affiliateWaiverApplied,
    delivery: deliveryFee,
    subtotal,
    discountApplied,
    quantityTierPct,
    multiMaterialSurcharge,
    total,
  };
}
